hail/python/hail/backend/service_backend.py

```diff
@@ -1,32 +1,43 @@
+from typing import Optional
 import os
 import aiohttp
 import json
+import warnings
 
 from hail.utils import FatalError
-from hail.expr.types import dtype
+from hail.expr.types import dtype, tvoid
 from hail.expr.table_type import ttable
 from hail.expr.matrix_type import tmatrix
 from hail.expr.blockmatrix_type import tblockmatrix
 
-from hailtop.config import get_deploy_config, get_user_config
+from hailtop.config import get_deploy_config, get_user_config, DeployConfig
 from hailtop.auth import service_auth_headers
-from hailtop.utils import async_to_blocking, retry_transient_errors
+from hailtop.utils import async_to_blocking, retry_transient_errors, secret_alnum_string, TransientError
 from hail.ir.renderer import CSERenderer
 
 from .backend import Backend
 from ..hail_logging import PythonOnlyLogger
+from ..fs.google_fs import GoogleCloudStorageFS
 
 
 class ServiceSocket:
-    def __init__(self, *, deploy_config=None):
+    def __init__(self, *, deploy_config: Optional[DeployConfig] = None):
         if not deploy_config:
             deploy_config = get_deploy_config()
+        self.deploy_config = deploy_config
         self.url = deploy_config.base_url('query')
-        self.session = aiohttp.ClientSession(headers=service_auth_headers(deploy_config, 'query'))
+        self._session: Optional[aiohttp.ClientSession] = None
+
+    async def session(self) -> aiohttp.ClientSession:
+        if self._session is None:
+            self._session = aiohttp.ClientSession(
+                headers=service_auth_headers(self.deploy_config, 'query'))
+        return self._session
 
     def close(self):
```

Comment:
So I think the right way to do this is to make `ServiceSocket` an (async) context manager.  Then you can construct the session in `__aenter__` and clean it up in `__aexit__`.

Reply:
I agree this class is kind of a mess, but I'm not sure a context manager helps? We're trying to share one Session for the entire lifetime of the service backend (many hail queries). If I change to a CM, I'd just __aenter__ it on line 78 / 99 in `class ServiceBackend`.